### PR TITLE
Add e2e test framework and port golden SQL tests

### PR DESCRIPTION
**Pull request by JamesGuthrie**
This change adds a dedicated e2e workspace which can be used to run
end-to-end tests against the promscale extension.

This change also also ports the golden SQL tests from the promscale
repository.

### CODE REVIEW

File: e2e/tests/golden-tests.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        node.get_host_port(5432).unwrap(),
+        DB
+    );
+
+    let mut client = postgres::Client::connect(connection_string, postgres::NoTls).unwrap();
+
+    let path = Path::new(resource);
+    assert!(path.exists());
+    let query = read_to_string(path).unwrap();
+    let result = client.simple_query(&query).unwrap();
```
Comment (JamesGuthrie) [inline e2e/tests/golden-tests.rs:45]: It would be nicer to just run `docker exec psql -c ...` and take the output of that. At the moment testcontainers doesn't support that, so I've taken the (annoyingly hacky) approach of running the file as a query against the DB and then transforming the results to a string.

The formatting of the final string could be improved to increase readability.

**Comment by sumerman** [comment]
Do you intend to use this to copy golden SQL tests from the promscale repo? If so, how do you plan to keep them up to date?

**Comment by JamesGuthrie** [comment]
> Do you intend to use this to copy and run golden SQL tests from the promscale repo? If so, how do you plan to keep them up to date?

Do you mean how do I intend to keep the golden sql tests here up to date with those from the promscale repo?

**Comment by sumerman** [comment]
> > Do you intend to use this to copy and run golden SQL tests from the promscale repo? If so, how do you plan to keep them up to date?
> 
> Do you mean how do I intend to keep the golden sql tests here up to date with those from the promscale repo?

I guess I was trying to ask two questions at once:
- do you intend to use this to run tests from the promscale repo?
- how do I intend to keep them to date with those from the promscale repo?

File: e2e/tests/golden-tests.rs (hunk 1 above)
Comment (cevian) [reply e2e/tests/golden-tests.rs:45]: Why don't we just do `docker exec psql -c ...` using process::Command::new("docker")... that's what testcontainers does anyway then 
```
let res = process::Command::new("docker")
            .arg("exec")
            .arg("-d")
            .arg(id)
            .arg("sh")
            .arg("-c")
            .arg(cmd)
            .stdout(Stdio::piped())
             .output()
             .expect("could not exec");
String::from_utf8_lossy(&output.stdout)

```

Comment (JamesGuthrie) [reply e2e/tests/golden-tests.rs]: Thanks for the input, I've done it this way.

File: Cargo.toml [hunk 2]
```diff
@@ -44,3 +47,6 @@ pgx-tests = "0.3.1"
 pgx = { git = "https://github.com/timescale/pgx", branch = "promscale-staging" }
 pgx-macros = { git = "https://github.com/timescale/pgx", branch = "promscale-staging" }
 pgx-tests = { git = "https://github.com/timescale/pgx", branch = "promscale-staging" }
+
+# from e2e workspace
+test-generator = { git = "https://github.com/JamesGuthrie/test-generator" }
```
Comment (sumerman) [inline Cargo.toml:52]: Why do we need it at the top level? Can't it live in the e2e workspace?

File: e2e/build.rs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+extern crate build_deps;
+
+fn main() {
+    build_deps::rerun_if_changed_paths("testdata/*.sql").unwrap();
```
Comment (sumerman) [inline e2e/build.rs:4]: Why is this necessary? To make `cargo test` re-run tests when SQL changes? Does it get cached otherwise?

File: e2e/tests/common/mod.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        PASSWORD,
+        node.get_host_port(5432).unwrap(),
+        DB
+    );
+
+    postgres::Client::connect(connection_string, postgres::NoTls).unwrap()
+}
+
+#[allow(dead_code)]
+pub fn exec_query(node: &Container<Cli, GenericImage>, query_path: &str) -> String {
```
Comment (sumerman) [inline e2e/tests/common/mod.rs:47]: A minor nit: it would be more clear to me if in the context of this function it was `sql_script` instead of `query`

File: e2e/tests/common/mod.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    let id = node.id();
+    let abs_query_path = "/".to_owned() + query_path;
+    let output = Command::new("docker")
```
Comment (sumerman) [inline e2e/tests/common/mod.rs:50]: It relies on `docker` command being available on the system, where the test runs, doesn't it?  I think this needs to be mentioned in the readme.

**Review by sumerman — APPROVED** [review]

File: Cargo.toml (hunk 2 above)
Comment (JamesGuthrie) [reply Cargo.toml:52]: https://doc.rust-lang.org/cargo/reference/workspaces.html#key-features

> The key points of workspaces are:
> - All packages share a common Cargo.lock file which resides in the workspace root.
> - All packages share a common [output directory](https://doc.rust-lang.org/cargo/guide/build-cache.html), which defaults to a directory named target in the workspace root.
> - **The [[patch]](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section), [[replace]](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-replace-section) and [[profile.*]](https://doc.rust-lang.org/cargo/reference/profiles.html) sections in Cargo.toml are only recognized in the root manifest, and ignored in member crates' manifests.**

File: e2e/build.rs (hunk 3 above)
Comment (JamesGuthrie) [reply e2e/build.rs:4]: This is necessary because the test cases are autogenerated from the `testdata` directory with a macro. If you make changes to the test code in the `src`, cargo knows that it needs to recompile. But without this `build.rs`, cargo doesn't know that it needs to recompile when you make changes to the `testdata` contents.

File: e2e/tests/common/mod.rs (hunk 4 above)
Comment (JamesGuthrie) [reply e2e/tests/common/mod.rs:47]: done
